server/raft.go

```diff
@@ -1160,13 +1161,38 @@ func (n *raft) isCatchingUp() bool {
 	return n.catchup != nil
 }
 
-// Lock should be held.
+// Lock should be held. This function may block for up to ~5ms to check
+// forward progress in some cases.
 func (n *raft) isCurrent() bool {
-	// First check if have had activity and we match commit and applied.
-	if n.commit == 0 || n.commit != n.applied {
-		n.debug("Not current, commit %d != applied %d", n.commit, n.applied)
-		return false
+	// First check if have had activity and that commit and applied are equal.
+	// If not, wait for a short period of time and see if we are making forward
+	// progress.
+	startCommit, startDelta := n.commit, n.commit-n.applied
+	if startCommit == 0 || startDelta > 0 {
+		var progress bool
+		for i := 0; i < 10; i++ { // 5ms, in 0.5ms increments
+			n.RUnlock()
+			time.Sleep(time.Millisecond / 2)
+			n.RLock()
+			curCommit, curDelta := n.commit, n.commit-n.applied
+			if startCommit == curCommit && curDelta < startDelta {
+				// Nothing new has come in but the gap is getting smaller, so
+				// we're making progress on the work we already had.
+				progress = true
+				break
+			}
+			if startCommit < curCommit && curDelta <= startDelta {
+				// New work has come in and the gap is either getting smaller
+				// or has stayed the same, so we're still making progress.
+				break
+			}
+		}
+		if !progress {
+			n.debug("Falling behind, commit %d != applied %d", n.commit, n.applied)
```

Comment:
Should this be a warn maybe?

Reply:
It was debug level before, if we want to change it to warn, is there more information we should include?

Reply:
Say it is inside health check logic, etc?

Reply:
Done :-)